parallel-task-set/src/lib.rs

```diff
@@ -0,0 +1,132 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::sync::Arc;
+use tokio::sync::Semaphore;
+use tokio::task::JoinSet;
+
+const DEFAULT_MAX_PARALLELISM: usize = 16;
+
+/// A collection of tokio tasks which execute in parallel on distinct tokio
+/// tasks, up to a user-specified maximum amount of parallelism.
+///
+/// This parallelism is achieved by spawning tasks on a [JoinSet],
+/// and may be further limited by the underlying machine's ability
+/// to execute many tokio tasks.
+///
+/// # Why not just use FuturesUnordered?
+///
+/// FuturesUnordered can execute any number of futures concurrently,
+/// but makes no attempt to execute them in parallel (assuming the underlying
+/// futures are not themselves spawning additional tasks).
+///
+/// # Why not just use a JoinSet?
+///
+/// The tokio [JoinSet] has not limit on the "maximum number of tasks".
+/// Given a bursty workload, it's possible to spawn an enormous number
+/// of tasks, which may not be desirable.
+///
+/// Although [ParallelTaskSet] uses a [JoinSet] internally, it also
+/// respects a parallism capacity.
+pub struct ParallelTaskSet<T> {
+    semaphore: Arc<Semaphore>,
+    set: JoinSet<T>,
```

Comment:
There *probably* should be a `Drop` implementation that calls `self.set.abort_all()`, so that the spawned tasks are torn down if the future creating the `ParallelTaskSet` is cancelled. 

Alternatively, if we're not convinced that this is correct in every case, this could expose a way for higher level code to decide about that (perhaps with a drop guard?).

Reply:
Isn't this the default behavior for a JoinSet? https://docs.rs/tokio/latest/src/tokio/task/join_set.rs.html#577-581

I can make it explicit too, but just wanna make sure I'm grokking what would be changing

Reply:
Ah, that must have changed at some point, I had gotten it into my head that it didn't do that. Nevermind!